### ISSUE TITLE
V4L2 driver - option to specify the intra refresh period (key frame rate/GoP)

### ISSUE THREAD

**Issue by zboobzor** [raspberrypi/linux#593]
If possible, could you please add an option to the V4L2 driver to specify the intra refresh period, like the --intra / -g option in raspivid.

Thanks.

**Comment by popcornmix** [comment]
@6by9 is this possible, or V4L2 doesn't have an API for?

**Comment by 6by9** [comment]
Looks like V4L2_CID_MPEG_VIDEO_H264_I_PERIOD is the correct parameter (search in http://hverkuil.home.xs4all.nl/spec/media.html for the description).
Easy enough to plumb in given half an hour or so.

**Comment by 6by9** [comment]
Done on my branch - https://github.com/6by9/linux/commit/7ce44a525504c84eb87e935eb70589047b6043cd

I blundered and pushed some other slightly non-functional code to that branch too :-( Need to fix that (or make a new branch) before I can make a pull request.

**Comment by zboobzor** [comment]
Thanks, just a few questions,
- Can the parameter be adjusted with v4l2-ctl ?
- Is the maximum value 1000 ?
  If so, could this restriction be removed or set higher ? (raspivid doesn't appear to enforce a maximum value)

**Comment by 6by9** [comment]
"v4l2-ctl --list-ctrls" prints out all the controls supported, which now includes:

```
h264_i_frame_period (int)    : min=1 max=1000 step=1 default=60 value=60
```

So "v4l2-ctl --set-ctrl=h264_i_frame_period=80 will set it for you via v4l2-ctl. 
This is true of any v4l2 parameter that the bit on the left from --list-ctrls is the string to use with --set-ctrl or --get-ctrl. 

I can't remember for certain, but I don't think our encoder allows this parameter to be changed whilst encoding. I'm happy to be proved wrong though.

Yes, I set the maximum at a fairly arbitrary 1000 - that's 33seconds at 30fps. You really want to go beyond that? I hope your transport level is 100% reliable then, as one dropped or corrupted frame and you'll be seeing the artifacts for a long time.
What maximum would you like to see? I could be convinced.

**Comment by zboobzor** [comment]
Thanks for the explanation about v4l2-ctl.

Concerning the maximum value, yes, my transport is 100% reliable, and I don't really want I-Frames at all (apart from the first one), so I'm currently sending an I-Frame once per hour with raspivid (my application never runs that long).
I understand that telling you I need a higher maximum value so I can abuse the option when I really need an I_FRAME_AT_START_THEN_NEVER_AGAIN option isn't very convincing :p but I assume that option doesn't exist.

**Comment by 6by9** [comment]
No problem. I'm just amazed that you really want no I-frames at all. It means seeking is near impossible in the stream, and is just "a little strange" ;-)

I'll double check what size the variables are down the stack and see if I can arrange >24hours for you. It literally is as simple as tweaking the number in 

```
+       V4L2_CID_MPEG_VIDEO_H264_I_PERIOD, MMAL_CONTROL_TYPE_STD,
+       1, 1000, 60, 1, NULL,
```

to increase the 1000 to something bigger. It has no side effects for anyone else.

(Never sending an i-frame again isn't currently possible - there is a check for I vs P frames that just does 

```
state->gop_counter--; 
if(!state->gop_counter) { 
   I-frame; 
   state->gop_counter=state->gop;
} 
else "P-frame" 
```

or equivalent. I don't really want to go tweaking in the codec code to change that).

**Comment by 6by9** [comment]
Actually you might be a happy bunny. From our internal codec API headers 

> If intraperiod
> is zero, all frames except for the first frame
> output will be non-intra frames.

and intraperiod referenced here is exactly the parameter we are setting. Sounds like you want me to change the minimum value too!

I haven't time to try it now, but I will try to do so in the next day or so.
The max value looks to be signed 32bit, which I make to be one I-frame every 2.26 years. Happy with that as a fall back (although I'm not testing it before releasing ;-) )

**Comment by 6by9** [comment]
Actually you might be a happy bunny. From our internal codec API headers 

> If intraperiod
> is zero, all frames except for the first frame
> output will be non-intra frames.

and intraperiod referenced here is exactly the parameter we are setting. Sounds like you want me to change the minimum value too!

I haven't time to try it now, but I will try to do so in the next day or so.

The max value looks to be signed 32bit, which I make to be one I-frame every 2.26 years. Are you happy with that as a fall back? (Although I'm not testing it for that long before releasing it - life's too short! ;-) )

**Comment by zboobzor** [comment]
The 0 setting sounds perfect :) but 2.26 years is definitely a decent fall back ;) (even if untested :p)

I do seem to remember trying strange values like 0 in raspivid with no success though.

## 

Just checked the raspivid code before posting and it looks like 0 is used as "not set", so not surprising it didn't work there.

**Comment by 6by9** [comment]
I've hassled Dom enough recently so haven't done a pull request, but I've pushed the changes to https://github.com/6by9/linux/commits/rpi-3.12.y and a prebuilt kernel module to https://github.com/6by9/RPiTest/blob/master/latest_v4l2_module/bcm2835-v4l2.ko

I did a quick hack to print out whether each buffer coming back was signaled as an I or P frame. With a value of 0, I got one I-frame in a stream of 3000 images. I'll take that as it working as documented.

Max is now 0x7FFFFFFF (2.26 years), so feel free to take that approach if necessary.
